File: golang/q800/q814.go
package q800

import . "github.com/zoroqi/stupid-self/golang"

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func pruneTree(root *TreeNode) *TreeNode {
	var dfs func(node *TreeNode) *TreeNode
	dfs = func(root *TreeNode) *TreeNode {
		if root == nil {
			return nil
		}
		root.Left = dfs(root.Left)
		root.Right = dfs(root.Right)
		if root.Right == nil && root.Left == nil && root.Val == 0 {
			return nil
		}
		return root
	}
	return dfs(root)
}

File: hs/src/Q800/Q814.hs
module Q800.Q814
(
    pruneTreePlanA
) where

import qualified Struct as Tree

pruneTreePlanA :: Integral a => Tree.BinTree a -> Tree.BinTree a
pruneTreePlanA Tree.Empty = Tree.Empty
pruneTreePlanA (Tree.Node a l r) =
                            case (nl, nr, a) of
                                (Tree.Empty, Tree.Empty, 0) -> Tree.Empty
                                (_, _, _) -> Tree.Node a nl nr
                            where
                                nl = pruneTreePlanA l
                                nr = pruneTreePlanA r
